plugins/fapolicyd.c

```diff
@@ -0,0 +1,175 @@
+#include "system.h"
+
+#include <rpm/rpmts.h>
+#include <rpm/rpmlog.h>
+#include "lib/rpmplugin.h"
+
+#include <fcntl.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+struct fapolicyd_data {
+    int fd;
+    int ready;
+    int first_scriptlet;
+    const char * fifo_path;
+};
+
+static struct fapolicyd_data fapolicyd_state = {
+    .fd = -1,
+    .ready = 0,
+    .first_scriptlet = 1,
+    .fifo_path = "/run/fapolicyd/fapolicyd.fifo",
+};
+
+static rpmRC open_fifo(struct fapolicyd_data* state)
+{
+    int fd = -1;
+    struct stat s;
+
+    fd = open(state->fifo_path, O_WRONLY);
+    if(fd == -1) {
+        rpmlog(RPMLOG_DEBUG, "Open: %s -> %s\n", state->fifo_path, strerror(errno));
+        return RPMRC_OK;
+    }
+
+    if (stat(state->fifo_path, &s) == -1) {
+        rpmlog(RPMLOG_DEBUG, "Stat: %s -> %s\n", state->fifo_path, strerror(errno));
+        close(fd);
+        return RPMRC_OK;
+    } else {
+
+        if (!S_ISFIFO(s.st_mode)) {
+            rpmlog(RPMLOG_DEBUG, "File: %s exists but it is not a pipe!\n", state->fifo_path);
+            close(fd);
+            return RPMRC_OK;
+        }
+
+        // we require pipe to have 0660 permissions
+        if (!(
+            (s.st_mode & S_IRUSR) &&
+            (s.st_mode & S_IWUSR) &&
+            !(s.st_mode & S_IXUSR) &&
+
+            (s.st_mode & S_IRGRP) &&
+            (s.st_mode & S_IWGRP) &&
+            !(s.st_mode & S_IXGRP) &&
+
+            !(s.st_mode & S_IROTH) &&
+            !(s.st_mode & S_IWOTH) &&
+            !(s.st_mode & S_IXOTH) )) {
+
+            rpmlog(RPMLOG_ERR, "File: %s has 0%d%d%d instead of 0660 \n",
+                    state->fifo_path,
+                    ((s.st_mode & S_IRUSR) ? 4 : 0) +
+                    ((s.st_mode & S_IWUSR) ? 2 : 0) +
+                    ((s.st_mode & S_IXUSR) ? 1 : 0),
+
+                    ((s.st_mode & S_IRGRP) ? 4 : 0) +
+                    ((s.st_mode & S_IWGRP) ? 2 : 0) +
+                    ((s.st_mode & S_IXGRP) ? 1 : 0),
+
+                    ((s.st_mode & S_IROTH) ? 4 : 0) +
+                    ((s.st_mode & S_IWOTH) ? 2 : 0) +
+                    ((s.st_mode & S_IXOTH) ? 1 : 0) );
+
+            close(fd);
+            return RPMRC_OK;
+        }
+    }
+
+    state->fd = fd;
+    state->ready = 1;
+    return RPMRC_OK;
+}
+
+static rpmRC write_fifo(struct fapolicyd_data* state, const char * str, size_t len)
+{
+    ssize_t ret = write(state->fd, str, len);
+
+    if (ret == -1) {
+        rpmlog(RPMLOG_DEBUG, "Write: %s -> %s\n", state->fifo_path, strerror(errno));
+    }
+
+    return RPMRC_OK;
+}
+
+static rpmRC fapolicyd_init(rpmPlugin plugin, rpmts ts)
+{
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST))
+        return RPMRC_OK;
+
+    open_fifo(&fapolicyd_state);
+    return RPMRC_OK;
+}
+
+static void fapolicyd_cleanup(rpmPlugin plugin)
+{
+    if (fapolicyd_state.ready)
+        close(fapolicyd_state.fd);
```

Comment:
Do conditionalize the close() on the fd itself, this ensures it gets closed no matter what. Or just leave the condition away, this is just a cleanup whose return we're not checking anyhow so no harm done calling close(-1).